classes/property_list.php

```
@@ -0,0 +1,346 @@
+     *
+     * @param string $key Key of element to update.
+     * @param mixed $value  Value to update element with.
+     */
+    public function update_element_value(string $key, $value) {
+        if (is_array($value)) {
+            debugging('property_list: Can\'t call update_element_value with array. Use update_element_array instead.',
+                    DEBUG_DEVELOPER);
+            return;
+        }
+        $this->plist_map( function($elvalue, $elkey, $parent) use ($key, $value) {
+            // Set new value.
+            if ($key === $elkey) {
+                $element = $parent->get($elkey);
+                // Limit update to boolean and strings types, and check value matches expected type.
+                if (($element instanceof CFString && is_string($value))
+                        || ($element instanceof CFNumber && is_numeric($value))
+                        || ($element instanceof CFBoolean && is_bool($value))) {
+                    $element->setValue($value);
+                } else {
+                    debugging('property_list: Can only update strings, numbers and bools with corresponding type.',
+                        DEBUG_DEVELOPER);
+                    return;
+                }
+            }
+        }, $this->cfpropertylist->getValue());
+    }
+
+    /**
+     * Update the array of any dict or array element with matching key.
+     *
+     * Will replace array.
+     *
+     * @param string $key Key of element to update.
+     * @param array $value  Array to update element with.
+     */
+    public function update_element_array(string $key, array $value) {
+        // Validate new array.
+        foreach ($value as $element) {
+            // If any element is not a CFType instance, then do nothing.
+            if (!($element instanceof CFType)) {
+                debugging('property_list: If updating an array in PList, it must only contain CFType objects.',
```

Need to throw an exception here.

---

See above